library/std/src/io/buffered.rs

```diff
@@ -816,6 +888,268 @@ impl<W> fmt::Display for IntoInnerError<W> {
+        let (lines, tail) = bufs.split_at(last_newline_buf_idx + 1);
+
+        // Write `lines` directly to the inner writer. In keeping with the
+        // `write` convention, make at most one attempt to add new (unbuffered)
+        // data. Because this write doesn't touch the BufWriter state directly,
+        // and the buffer is known to be empty, we don't need to worry about
+        // self.panicked here.
+        let flushed = self.inner_mut().write_vectored(lines)?;
+
+        // If inner returns Ok(0), propagate that to the caller without
+        // doing additional buffering; otherwise we're just guaranteeing
+        // an "ErrorKind::WriteZero" later.
+        if flushed == 0 {
+            return Ok(0);
+        }
+
+        // Don't try to reconstruct the exact amount written; just bail
+        // in the event of a partial write
+        let lines_len = lines.iter().map(|buf| buf.len()).sum();
+        if flushed < lines_len {
+            return Ok(flushed);
+        }
+
+        // Now that the write has succeeded, buffer the rest (or as much of the
+        // rest as possible)
+        let buffered: usize = tail
+            .iter()
+            .filter(|buf| !buf.is_empty())
+            .map(|buf| self.buffer.write_to_buf(buf))
+            .take_while(|&n| n > 0)
+            .sum();
+
+        Ok(flushed + buffered)
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        self.buffer.is_write_vectored()
+    }
+
+    /// Write some data into this BufReader with line buffering. This means
+    /// that, if any newlines are present in the data, the data up to the last
+    /// newline is sent directly to the underlying writer, and data after it
+    /// is buffered.
+    ///
+    /// Because this function attempts to send completed lines to the underlying
+    /// writer, it will also flush the existing buffer if it contains any
+    /// newlines, even if the incoming data does not contain any newlines.
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        let newline_idx = match memchr::memrchr(b'\n', buf) {
+            // If there are no new newlines (that is, if this write is less than
+            // one line), just do a regular buffered write
+            None => {
+                self.flush_if_completed_line()?;
+                return self.buffer.write_all(buf);
+            }
+            // Otherwise, arrange for the lines to be written directly to the
+            // inner writer.
+            Some(newline_idx) => newline_idx,
+        };
+
+        // Flush existing content to prepare for our write
+        self.buffer.flush_buf()?;
+
+        // This is what we're going to try to write directly to the inner
+        // writer. The rest will be buffered, if nothing goes wrong.
+        let (lines, tail) = buf.split_at(newline_idx + 1);
+
+        // Write `lines` directly to the inner writer, bypassing the buffer.
+        self.inner_mut().write_all(lines)?;
```

Comment:
When using `println!`, one desirable property is that the formatted output is
written using a single syscall. To take advantage of write atomicity, and avoid
interleaving the output with that from other processes that might be sharing
the same stdout.

The implementation here performs two syscalls whenever there are any formatting
operations in `println!`. The presence of terminal `\n` triggers the flush of
existing buffer content and then the final string with a newline is flushed
separately.

For example, `println!("Hello {}!", "world");`, leads to:

```
write(1, "Hello world", 11)             = 11
write(1, "!\n", 2)                      = 2
```

Reply:
Hmm. It's a good point. The problem is this:

Suppose that, as a solution (in your example), we buffer the second print, so that we can do the whole thing in a single write. At that point, we are *obligated* to return an Ok(2), because we have "successfully" written 2 bytes (by adding them to the buffer). This means that we can't actually attempt to send the byte to the underlying device, because if the device returns an error, we'd have to ignore that error (or save it and return it on the *next* write), which (along with extra, unnecessary data buffering) is something I was trying to avoid.

One workaround to this would be to buffer the data, attempt a write, and if the write returns an error, remove that data from the buffer. However, in that case, if the first write returns Ok(12) (representing "Hello World") and the second write returns an error, there's no way to report a consistent result to the caller without also suppressing that error.

That being said, this problem probably *can* be addressed in the `write_all` specialization (used by default `write_fmt`), since it's not necessary to try to preserve a recoverable / knowable state in the event of an error there. I'll see about retooling `write_all` a bit to address this concern.

Reply:
Thanks for the feedback! I've updated `write_all` to address this concern, and added a test verifying the behaivor.

Reply:
Thanks for addressing this! I expected that doing so might be challenging in
context of other goals. Fortunately, the fact that formatting goes through
`write_all` made it quite manageable. Nice!